### PR TITLE
Add possibility of configuration to Environment Listeners

### LINKED ISSUE #829: Add mode to SystemEnvironmentProjectListener etc
The mode should allow us to control when the env is added - always or only if it doesn't already exist.

`SystemEnvironmentProjectListener("ENV_NAME", "test", Mode.OVERWRITE)`

vs

`SystemEnvironmentProjectListener("ENV_NAME", "test", Mode.CREATE)`

### PR DESCRIPTION
**Pull request by LeoColman**
Closes #829

### CODE REVIEW

**Comment by LeoColman** [comment]
I think that I might be missing some tests here, but I'm not sure

File: kotlintest-extensions/src/main/kotlin/io/kotlintest/extensions/system/MapOverrides.kt [hunk 1]
```diff
@@ -1,12 +1,34 @@
 package io.kotlintest.extensions.system
 
+
+enum class SystemOverrideMode { ALLOW_OVERRIDE, DENY_OVERRIDE }
```
Comment (sksamuel) [inline kotlintest-extensions/src/main/kotlin/io/kotlintest/extensions/system/MapOverrides.kt]: The two possibilities, actually lets say three.

If already set ignore
If already set override
Error if already set

So maybe the nomenclature can be:

- SET_OR_OVERRIDE
- SET_OR_ERROR (default)
- SET_OR_IGNORE

?

**Review by sksamuel — CHANGES_REQUESTED** [review]
As usual from your PRs it's great, but I think we should allow all three options now.

File: kotlintest-extensions/src/main/kotlin/io/kotlintest/extensions/system/MapOverrides.kt (hunk 1 above)
Comment (LeoColman) [reply kotlintest-extensions/src/main/kotlin/io/kotlintest/extensions/system/MapOverrides.kt]: My only point to this is that `SET_OR_OVERRIDE` should be set as default. It's the default now, and I don't think we need to change that behavior and potentially break clients

**Comment by sksamuel** [comment]
Agreed.

File: kotlintest-extensions/src/main/kotlin/io/kotlintest/extensions/system/MapOverrides.kt (hunk 1 above)
Comment (sksamuel) [reply kotlintest-extensions/src/main/kotlin/io/kotlintest/extensions/system/MapOverrides.kt]: Yes ok agreed.

**Comment by LeoColman** [comment]
@sksamuel I think it's done.

Please review again!

**Comment by LeoColman** [comment]
I decided that setting the default to `SetOrError` is better. It should be the expected, and if clients' code break, it is a good thing

**Comment by sksamuel** [comment]
Do you want me to finish this off?

**Review by sksamuel — APPROVED** [review]

**Comment by LeoColman** [comment]
Did I miss anything?

**Comment by sksamuel** [comment]
Don't think so

**Pull request merged**